aws/resource_aws_neptune_subnet_group_test.go

```diff
@@ -0,0 +1,385 @@
+package aws
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/neptune"
+)
+
+func TestAccAWSNeptuneSubnetGroup_basic(t *testing.T) {
+	var v neptune.DBSubnetGroup
+
+	testCheck := func(*terraform.State) error {
+		return nil
+	}
+
+	rName := fmt.Sprintf("tf-test-%d", acctest.RandInt())
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckNeptuneSubnetGroupDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccNeptuneSubnetGroupConfig(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckNeptuneSubnetGroupExists(
+						"aws_neptune_subnet_group.foo", &v),
+					resource.TestCheckResourceAttr(
+						"aws_neptune_subnet_group.foo", "name", rName),
+					resource.TestCheckResourceAttr(
+						"aws_neptune_subnet_group.foo", "description", "Managed by Terraform"),
+					testCheck,
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSNeptuneSubnetGroup_namePrefix(t *testing.T) {
+	var v neptune.DBSubnetGroup
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckNeptuneSubnetGroupDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccNeptuneSubnetGroupConfig_namePrefix,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckNeptuneSubnetGroupExists(
+						"aws_neptune_subnet_group.test", &v),
+					resource.TestMatchResourceAttr(
+						"aws_neptune_subnet_group.test", "name", regexp.MustCompile("^tf_test-")),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSNeptuneSubnetGroup_generatedName(t *testing.T) {
+	var v neptune.DBSubnetGroup
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckNeptuneSubnetGroupDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccNeptuneSubnetGroupConfig_generatedName,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckNeptuneSubnetGroupExists(
+						"aws_neptune_subnet_group.test", &v),
+				),
+			},
+		},
+	})
+}
+
+// Regression test for https://github.com/hashicorp/terraform/issues/2603 and
+// https://github.com/hashicorp/terraform/issues/2664
+func TestAccAWSNeptuneSubnetGroup_withUndocumentedCharacters(t *testing.T) {
```

Comment:
Acceptable characters are in the [CreateDBSubnetGroup API documentation](https://docs.aws.amazon.com/neptune/latest/userguide/API_CreateDBSubnetGroup.html):

> Constraints: Must contain no more than 255 letters, numbers, periods, underscores, spaces, or hyphens. Must not be default.

This validation should be performed in unit testing and not acceptance testing.

Reply:
removed this acceptance test.

Reply:
Unit testing for this validation already covered in `validators_test.go`. still we need to add unit testing here?